In [1]:
import pandas as pd

data = pd.read_csv("E:/Data Science/CSV dataset/melb_data.csv")
data

Suburb           Address  Rooms Type      Price Method  \
0         Abbotsford      85 Turner St      2    h  1480000.0      S   
1         Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2         Abbotsford      5 Charles St      3    h  1465000.0     SP   
3         Abbotsford  40 Federation La      3    h   850000.0     PI   
4         Abbotsford       55a Park St      4    h  1600000.0     VB   
...              ...               ...    ...  ...        ...    ...   
13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  \
0        Biggin   3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0   
1        Biggin   4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0   
2        Biggin   4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0   
3        Biggin   4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0   
4        Nelson   4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0   
...         ...         ...       ...       ...  ...       ...  ...       ...   
13575     Barry  26/08/2017      16.7    3150.0  ...       2.0  2.0     652.0   
13576  Williams  26/08/2017       6.8    3016.0  ...       2.0  2.0     333.0   
13577     Raine  26/08/2017       6.8    3016.0  ...       2.0  4.0     436.0   
13578   Sweeney  26/08/2017       6.8    3016.0  ...       1.0  5.0     866.0   
13579   Village  26/08/2017       6.3    3013.0  ...       1.0  1.0     362.0   

       BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0               NaN        NaN        Yarra -37.79960   144.99840   
1              79.0     1900.0        Yarra -37.80790   144.99340   
2             150.0     1900.0        Yarra -37.80930   144.99440   
3               NaN        NaN        Yarra -37.79690   144.99690   
4             142.0     2014.0        Yarra -37.80720   144.99410   
...             ...        ...          ...       ...         ...   
13575           NaN     1981.0          NaN -37.90562   145.16761   
13576         133.0     1995.0          NaN -37.85927   144.87904   
13577           NaN     1997.0          NaN -37.85274   144.88738   
13578         157.0     1920.0          NaN -37.85908   144.89299   
13579         112.0     1920.0          NaN -37.81188   144.88449   

                       Regionname Propertycount  
0           Northern Metropolitan        4019.0  
1           Northern Metropolitan        4019.0  
2           Northern Metropolitan        4019.0  
3           Northern Metropolitan        4019.0  
4           Northern Metropolitan        4019.0  
...                           ...           ...  
13575  South-Eastern Metropolitan        7392.0  
13576        Western Metropolitan        6380.0  
13577        Western Metropolitan        6380.0  
13578        Western Metropolitan        6380.0  
13579        Western Metropolitan        6543.0  

[13580 rows x 21 columns]

In [2]:
## Extracting the columns

data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [3]:
## Dropping missing values in the dataset using the dropna() function

data.dropna(axis = 0)                   ## axis = 0 means selecting the rows only

Suburb          Address  Rooms Type      Price Method  \
1        Abbotsford  25 Bloomburg St      2    h  1035000.0      S   
2        Abbotsford     5 Charles St      3    h  1465000.0     SP   
4        Abbotsford      55a Park St      4    h  1600000.0     VB   
6        Abbotsford     124 Yarra St      3    h  1876000.0      S   
7        Abbotsford    98 Charles St      2    h  1636000.0      S   
...             ...              ...    ...  ...        ...    ...   
12205    Whittlesea    30 Sherwin St      3    h   601000.0      S   
12206  Williamstown      75 Cecil St      3    h  1050000.0     VB   
12207  Williamstown    2/29 Dover Rd      1    u   385000.0     SP   
12209       Windsor  201/152 Peel St      2    u   560000.0     PI   
12212    Yarraville  54 Pentland Pde      6    h  2450000.0     VB   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
1             Biggin   4/02/2016       2.5    3067.0  ...       1.0  0.0   
2             Biggin   4/03/2017       2.5    3067.0  ...       2.0  0.0   
4             Nelson   4/06/2016       2.5    3067.0  ...       1.0  2.0   
6             Nelson   7/05/2016       2.5    3067.0  ...       2.0  0.0   
7             Nelson   8/10/2016       2.5    3067.0  ...       1.0  2.0   
...              ...         ...       ...       ...  ...       ...  ...   
12205            Ray  29/07/2017      35.5    3757.0  ...       2.0  1.0   
12206       Williams  29/07/2017       6.8    3016.0  ...       1.0  0.0   
12207       Williams  29/07/2017       6.8    3016.0  ...       1.0  1.0   
12209  hockingstuart  29/07/2017       4.6    3181.0  ...       1.0  1.0   
12212        Village  29/07/2017       6.3    3013.0  ...       3.0  2.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
1         156.0         79.00     1900.0        Yarra -37.80790   144.99340   
2         134.0        150.00     1900.0        Yarra -37.80930   144.99440   
4         120.0        142.00     2014.0        Yarra -37.80720   144.99410   
6         245.0        210.00     1910.0        Yarra -37.80240   144.99930   
7         256.0        107.00     1890.0        Yarra -37.80600   144.99540   
...         ...           ...        ...          ...       ...         ...   
12205     972.0        149.00     1996.0   Whittlesea -37.51232   145.13282   
12206     179.0        115.00     1890.0  Hobsons Bay -37.86558   144.90474   
12207       0.0         35.64     1967.0  Hobsons Bay -37.85588   144.89936   
12209       0.0         61.60     2012.0  Stonnington -37.85581   144.99025   
12212    1087.0        388.50     1920.0  Maribyrnong -37.81038   144.89389   

                  Regionname Propertycount  
1      Northern Metropolitan        4019.0  
2      Northern Metropolitan        4019.0  
4      Northern Metropolitan        4019.0  
6      Northern Metropolitan        4019.0  
7      Northern Metropolitan        4019.0  
...                      ...           ...  
12205      Northern Victoria        2170.0  
12206   Western Metropolitan        6380.0  
12207   Western Metropolitan        6380.0  
12209  Southern Metropolitan        4380.0  
12212   Western Metropolitan        6543.0  

[6196 rows x 21 columns]

Selecting The Prediction Target

You can pull out a variable with dot-notation. 
This single column is stored in a Series, which is broadly like a DataFrame with only a single column of data.

We'll use the dot notation to select the column we want to predict, which is called the prediction target. By convention, the prediction target is called y.

In [4]:
## Selecting the prediction variable here..

y = data.Price

Choosing "Features"

The columns that are inputted into our model (and later used to make predictions) are called "features." 
In this case, those would be the columns used to determine the home price. Sometimes, all columns except the target are used as features or sometimes with fewer features.

We select multiple features by providing a list of column names inside brackets. Each item in that list should be a string (with quotes).

In [5]:
features = ['Rooms', 'Bathroom', 'Landsize', 'Landsize', 'Longtitude']

In [6]:
# assigning the features to a variable

x = data[features]
print(x)

       Rooms  Bathroom  Landsize  Landsize  Longtitude
0          2       1.0     202.0     202.0   144.99840
1          2       1.0     156.0     156.0   144.99340
2          3       2.0     134.0     134.0   144.99440
3          3       2.0      94.0      94.0   144.99690
4          4       1.0     120.0     120.0   144.99410
...      ...       ...       ...       ...         ...
13575      4       2.0     652.0     652.0   145.16761
13576      3       2.0     333.0     333.0   144.87904
13577      3       2.0     436.0     436.0   144.88738
13578      4       1.0     866.0     866.0   144.89299
13579      4       1.0     362.0     362.0   144.88449

[13580 rows x 5 columns]


In [7]:
## Getting the stats of the variables

x.describe()

Rooms      Bathroom       Landsize       Landsize    Longtitude
count  13580.000000  13580.000000   13580.000000   13580.000000  13580.000000
mean       2.937997      1.534242     558.416127     558.416127    144.995216
std        0.955748      0.691712    3990.669241    3990.669241      0.103916
min        1.000000      0.000000       0.000000       0.000000    144.431810
25%        2.000000      1.000000     177.000000     177.000000    144.929600
50%        3.000000      1.000000     440.000000     440.000000    145.000100
75%        3.000000      2.000000     651.000000     651.000000    145.058305
max       10.000000      8.000000  433014.000000  433014.000000    145.526350

In [8]:
## Checking for NaN values in the input variable x

x.isnull().sum()

Rooms         0
Bathroom      0
Landsize      0
Landsize      0
Longtitude    0
dtype: int64

The steps to building and using a model are:

Define: What type of model will it be? A decision tree? Some other type of model? Some other parameters of the model type are specified too.

Fit: Capture patterns from provided data. This is the heart of modeling.

Predict: Just what it sounds like

Evaluate: Determine how accurate the model's predictions are 

In [9]:
from sklearn.tree import DecisionTreeRegressor

melbourne_model = DecisionTreeRegressor(random_state = 1)

# fitting the Model

melbourne_model.fit(x, y)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=1, splitter='best')

Many machine learning models allow some randomness in model training. Specifying a number for random_state ensures you get the same results in each run. This is considered a good practice. You use any number, and model quality won't depend meaningfully on exactly what value you choose.

We now have a fitted model that we can use to make predictions.

In practice, you'll want to make predictions for new houses coming on the market rather than the houses we already have prices for. But we'll make predictions for the first few rows of the training data to see how the predict function works.

In [10]:
print('Making predictions for 5 new houses\n')
print(x.head())
print('\n\nThe predictions are:: \n')
print(melbourne_model.predict(x.head()))

Making predictions for 5 new houses

   Rooms  Bathroom  Landsize  Landsize  Longtitude
0      2       1.0     202.0     202.0    144.9984
1      2       1.0     156.0     156.0    144.9934
2      3       2.0     134.0     134.0    144.9944
3      3       2.0      94.0      94.0    144.9969
4      4       1.0     120.0     120.0    144.9941


The predictions are:: 

[1480000. 1035000. 1465000.  850000. 1600000.]


In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         13580 non-null  object 
 1   Address        13580 non-null  object 
 2   Rooms          13580 non-null  int64  
 3   Type           13580 non-null  object 
 4   Price          13580 non-null  float64
 5   Method         13580 non-null  object 
 6   SellerG        13580 non-null  object 
 7   Date           13580 non-null  object 
 8   Distance       13580 non-null  float64
 9   Postcode       13580 non-null  float64
 10  Bedroom2       13580 non-null  float64
 11  Bathroom       13580 non-null  float64
 12  Car            13518 non-null  float64
 13  Landsize       13580 non-null  float64
 14  BuildingArea   7130 non-null   float64
 15  YearBuilt      8205 non-null   float64
 16  CouncilArea    12211 non-null  object 
 17  Lattitude      13580 non-null  float64
 18  Longti

In [12]:
## Selecting other columns to predict the price

features2 = ['Rooms', 'Distance', 'Bedroom2', 'Bathroom', 'Landsize', 'Lattitude', 'Longtitude', 'Propertycount']

X = data[features2]
print(X)

# Selecting the target variable

Y = data['Price']


       Rooms  Distance  Bedroom2  Bathroom  Landsize  Lattitude  Longtitude  \
0          2       2.5       2.0       1.0     202.0  -37.79960   144.99840   
1          2       2.5       2.0       1.0     156.0  -37.80790   144.99340   
2          3       2.5       3.0       2.0     134.0  -37.80930   144.99440   
3          3       2.5       3.0       2.0      94.0  -37.79690   144.99690   
4          4       2.5       3.0       1.0     120.0  -37.80720   144.99410   
...      ...       ...       ...       ...       ...        ...         ...   
13575      4      16.7       4.0       2.0     652.0  -37.90562   145.16761   
13576      3       6.8       3.0       2.0     333.0  -37.85927   144.87904   
13577      3       6.8       3.0       2.0     436.0  -37.85274   144.88738   
13578      4       6.8       4.0       1.0     866.0  -37.85908   144.89299   
13579      4       6.3       4.0       1.0     362.0  -37.81188   144.88449   

       Propertycount  
0             4019.0  
1    

In [13]:
## Training our model

from sklearn.tree import DecisionTreeRegressor

melbourne_model2 = DecisionTreeRegressor(random_state = 1)

# Fitting our model 

melbourne_model2.fit(X, Y)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=1, splitter='best')

In [14]:
## Making the predictions 

print('5 New houses stats :: \n')
print(X.head())

print('\nPredictions are:: \n')
print(melbourne_model2.predict(X.head()))

5 New houses stats :: 

   Rooms  Distance  Bedroom2  Bathroom  Landsize  Lattitude  Longtitude  \
0      2       2.5       2.0       1.0     202.0   -37.7996    144.9984   
1      2       2.5       2.0       1.0     156.0   -37.8079    144.9934   
2      3       2.5       3.0       2.0     134.0   -37.8093    144.9944   
3      3       2.5       3.0       2.0      94.0   -37.7969    144.9969   
4      4       2.5       3.0       1.0     120.0   -37.8072    144.9941   

   Propertycount  
0         4019.0  
1         4019.0  
2         4019.0  
3         4019.0  
4         4019.0  

Predictions are:: 

[1480000. 1035000. 1465000.  850000. 1600000.]


You'd first need to summarize the model quality into an understandable way. If you compare predicted and actual home values for 10,000 houses, you'll likely find mix of good and bad predictions. Looking through a list of 10,000 predicted and actual values would be pointless. We need to summarize this into a single metric.

There are many metrics for summarizing model quality, but we'll start with one called Mean Absolute Error (also called MAE). 

The prediction error for each house is:     error = actual − predicted

So, if a house cost $150,000 and you predicted it would cost $100,000 the error is $50,000.

With the MAE metric, we take the absolute value of each error. This converts each error to a positive number. We then take the average of those absolute errors. This is our measure of model quality. In plain English, it can be said as

To calculate MAE, we first need a model. That is built in a hidden cell below, which you can review by clicking the code button.

Once we have a model, here is how we calculate the mean absolute error:


In [15]:
from sklearn.metrics import mean_absolute_error

predicted_home_prices = melbourne_model2.predict(X)

mean_absolute_error(predicted_home_prices, Y)

978.3714285714284

The Problem with "In-Sample" Scores

The measure we just computed can be called an "in-sample" score. We used a single "sample" of houses for both building the model and evaluating it. Here's why this is bad.

Imagine that, in the large real estate market, door color is unrelated to home price.

However, in the sample of data you used to build the model, all homes with green doors were very expensive. The model's job is to find patterns that predict home prices, so it will see this pattern, and it will always predict high prices for homes with green doors.

Since this pattern was derived from the training data, the model will appear accurate in the training data.

But if this pattern doesn't hold when the model sees new data, the model would be very inaccurate when used in practice.

Since models' practical value come from making predictions on new data, we measure performance on data that wasn't used to build the model. The most straightforward way to do this is to exclude some data from the model-building process, and then use those to test the model's accuracy on data it hasn't seen before. This data is called validation data.

Coding It :::

The scikit-learn library has a function train_test_split to break up the data into two pieces. We'll use some of that data as training data to fit the model, and we'll use the other data as validation data to calculate mean_absolute_error.

In [16]:
from sklearn.model_selection import train_test_split

# split data into training and validation data, for both features and target
# The split is based on a random number generator. Supplying a numeric value to
# the random_state argument guarantees we get the same split every time we
# run this script.

train_x, val_x, train_y, val_y = train_test_split(X, Y , random_state = 1)

# Defining the model again

melbourne_model3 = DecisionTreeRegressor()

# fitting the model
melbourne_model3.fit(train_x, train_y)

# Calculating predicted prices on validation values

predicted_home_prices2 = melbourne_model3.predict(val_x)

## Calculating the mean absolute error 


print("Mean absoslute error of the model:: \n ")
mean_absolute_error(val_y, predicted_home_prices2)



Mean absoslute error of the model:: 
 


238350.58949435444

Wow!

Your mean absolute error for the in-sample data was about 500 dollars. Out-of-sample it is more than 250,000 dollars.

This is the difference between a model that is almost exactly right, and one that is unusable for most practical purposes. As a point of reference, the average home value in the validation data is 1.1 million dollars. So the error in new data is about a quarter of the average home value.

This is the difference between a model that is almost exactly right, and one that is unusable for most practical purposes. As a point of reference, the average home value in the validation data is 1.1 million dollars. So the error in new data is about a quarter of the average home value.



In practice, it's not uncommon for a tree to have 10 splits between the top level (all houses) and a leaf. As the tree gets deeper, the dataset gets sliced up into leaves with fewer houses. If a tree only had 1 split, it divides the data into 2 groups. If each group is split again, we would get 4 groups of houses. If we keep doubling the number of groups by adding more splits at each level, we'll have 210 groups of houses by the time we get to the 10th level. That's 1024 leaves.

When we divide the houses amongst many leaves, we also have fewer houses in each leaf. Leaves with very few houses will make predictions that are quite close to those homes' actual values, but they may make very unreliable predictions for new data (because each prediction is based on only a few houses).

This is a phenomenon called OVERFITTING, where a model matches the training data almost perfectly, but does poorly in validation and other new data. 

If a tree divides houses into only 2 or 4, each group still has a wide variety of houses. Resulting predictions may be far off for most houses. When a model fails to capture important distinctions and patterns in the data, so it performs poorly even in training data, that is called UNDERFITTING.



Example

There are a few alternatives for controlling the tree depth. But the max_leaf_nodes argument provides a very sensible way to control overfitting vs underfitting. The more leaves we allow the model to make, the more we move from the underfitting area in the above graph to the overfitting area.

We can use a utility function to help compare MAE scores from different values for max_leaf_nodes:

In [17]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error


def max_leaves(max_leaves_nodes ,train_x, val_x, train_y, val_y):
        """
        Function to limit the max leaves in the model
        """
        model = DecisionTreeRegressor(max_leaf_nodes = max_leaves_nodes, random_state = 0)
        model.fit(train_x, train_y)
        pred_val = model.predict(val_x)
        mae = mean_absolute_error(val_y, pred_val)
        return mae
    

# compare MAE with differing values of max_leaf_nodes

for max_leaves_nodes in [5,50,500,5000]:
    my_mae = max_leaves(max_leaves_nodes , train_x, val_x, train_y, val_y)
    print('Max leaves nodes: %d \t\t\tMean Absolute Error is:%d'%(max_leaves_nodes, my_mae))
    

Max leaves nodes: 5 			Mean Absolute Error is:364815
Max leaves nodes: 50 			Mean Absolute Error is:255298
Max leaves nodes: 500 			Mean Absolute Error is:215505
Max leaves nodes: 5000 			Mean Absolute Error is:234638


Conclusion

Here's the takeaway: Models can suffer from either:

Overfitting: capturing spurious patterns that won't recur in the future, leading to less accurate predictions, or
Underfitting: failing to capture relevant patterns, again leading to less accurate predictions.

We use validation data, which isn't used in model training, to measure a candidate model's accuracy. 

In [18]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

forest_model = RandomForestRegressor(random_state = 1)
forest_model.fit(train_x, train_y)
forest_pred = forest_model.predict(val_x)
mean_abs_error = mean_absolute_error(val_y, forest_pred)
print('Mean absolute error of Random Forest model is:: ', mean_abs_error)

Mean absolute error of Random Forest model is::  174631.5377777778


Dealing with missing values::

Three Approaches:: 

1) A Simple Option: Drop Columns with Missing Values

Unless most values in the dropped columns are missing, the model loses access to a lot of (potentially useful!) information with this approach. As an extreme example, consider a dataset with 10,000 rows, where one important column is missing a single entry. This approach would drop the column entirely!

2) A Better Option: Imputation

Imputation fills in the missing values with some number. For instance, we can fill in the mean value along each column.
The imputed value won't be exactly right in most cases, but it usually leads to more accurate models than you would get from dropping the column entirely.

3) An Extension To Imputation

Imputation is the standard approach, and it usually works well. However, imputed values may be systematically above or below their actual values (which weren't collected in the dataset). Or rows with missing values may be unique in some other way. In that case, your model would make better predictions by considering which values were originally missing.

we use SimpleImputer to replace missing values with the mean value along each column.

Although it's simple, filling in the mean value generally performs quite well (but this varies by dataset). While statisticians have experimented with more complex ways to determine imputed values (such as regression imputation, for instance), the complex strategies typically give no additional benefit once you plug the results into sophisticated machine learning models.

In [21]:
data.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom            0
Car                62
Landsize            0
BuildingArea     6450
YearBuilt        5375
CouncilArea      1369
Lattitude           0
Longtitude          0
Regionname          0
Propertycount       0
dtype: int64

ABove result shows few columns have lots of missing values.

In [34]:
data[data.Car.isnull()]

Suburb            Address  Rooms Type      Price Method  \
12221       Ascot Vale     132 The Parade      3    h   985000.0      S   
12247   Brunswick East        18 Ethel St      2    h  1023000.0      S   
12259     Clifton Hill      34 Fenwick St      3    h  1436000.0      S   
12320    Glen Waverley      19 Diamond Av      3    h  1370000.0      S   
12362          Newport  11 Collingwood Rd      4    h  1180000.0     PI   
...                ...                ...    ...  ...        ...    ...   
13496     Moonee Ponds     46 Eglinton St      4    h  1525000.0      S   
13508  North Melbourne       9 Erskine St      2    h  1080000.0      S   
13522   Port Melbourne      201 Stokes St      2    h  1515000.0     SP   
13524          Prahran   17 Packington Pl      2    h  1365000.0      S   
13550   Sunshine North     5 Berkshire Rd      3    h   595000.0     SP   

         SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
12221       Brad   3/09/2017       4.3    3032.0  ...       1.0  NaN   
12247     Domain   3/09/2017       4.0    3057.0  ...       1.0  NaN   
12259     Jellis   3/09/2017       3.6    3068.0  ...       2.0  NaN   
12320  Fletchers   3/09/2017      16.7    3150.0  ...       1.0  NaN   
12362   Williams   3/09/2017       6.2    3015.0  ...       1.0  NaN   
...          ...         ...       ...       ...  ...       ...  ...   
13496     Nelson  26/08/2017       6.2    3039.0  ...       3.0  NaN   
13508     Jellis  26/08/2017       1.8    3051.0  ...       1.0  NaN   
13522   Marshall  26/08/2017       3.5    3207.0  ...       2.0  NaN   
13524     Jellis  26/08/2017       4.6    3181.0  ...       1.0  NaN   
13550      Barry  26/08/2017      10.5    3020.0  ...       1.0  NaN   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
12221     245.0          91.0     1945.0          NaN -37.77215   144.91144   
12247     154.0          76.0     1890.0          NaN -37.77221   144.97537   
12259     123.0         128.0     1990.0          NaN -37.78888   145.00036   
12320     652.0           NaN        NaN          NaN -37.87170   145.17267   
12362     545.0           NaN        NaN          NaN -37.84399   144.89125   
...         ...           ...        ...          ...       ...         ...   
13496     233.0           NaN        NaN          NaN -37.76884   144.91708   
13508     100.0          67.0     1890.0          NaN -37.79524   144.94642   
13522     197.0           NaN        NaN          NaN -37.83754   144.93954   
13524     206.0         100.0     1900.0          NaN -37.85569   145.00522   
13550     589.0           NaN        NaN          NaN -37.76599   144.84445   

                  Regionname Propertycount  
12221   Western Metropolitan        6567.0  
12247  Northern Metropolitan        5533.0  
12259  Northern Metropolitan        2954.0  
12320   Eastern Metropolitan       15321.0  
12362   Western Metropolitan        5498.0  
...                      ...           ...  
13496   Western Metropolitan        6232.0  
13508  Northern Metropolitan        6821.0  
13522  Southern Metropolitan        8648.0  
13524  Southern Metropolitan        7717.0  
13550   Western Metropolitan        4217.0  

[62 rows x 21 columns]

In [35]:
## Dropping all rows which have NaN values 

data.dropna(how = 'any').shape            # How = any selects all rows which have NaN values.


(6196, 21)

In [36]:
data.dropna(how = 'all').shape          # How = all selects any row which have all NaN values not just one.

(13580, 21)

In [37]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         13580 non-null  object 
 1   Address        13580 non-null  object 
 2   Rooms          13580 non-null  int64  
 3   Type           13580 non-null  object 
 4   Price          13580 non-null  float64
 5   Method         13580 non-null  object 
 6   SellerG        13580 non-null  object 
 7   Date           13580 non-null  object 
 8   Distance       13580 non-null  float64
 9   Postcode       13580 non-null  float64
 10  Bedroom2       13580 non-null  float64
 11  Bathroom       13580 non-null  float64
 12  Car            13518 non-null  float64
 13  Landsize       13580 non-null  float64
 14  BuildingArea   7130 non-null   float64
 15  YearBuilt      8205 non-null   float64
 16  CouncilArea    12211 non-null  object 
 17  Lattitude      13580 non-null  float64
 18  Longti

In [41]:
train_x.head()

Rooms  Distance  Bedroom2  Bathroom  Landsize  Lattitude  Longtitude  \
7623       2       4.5       2.0       1.0    1556.0  -37.77210   144.97190   
776        6      13.0       6.0       3.0     708.0  -37.91810   145.04400   
5684       1       3.3       1.0       1.0       0.0  -37.84270   145.00060   
12888      4       3.0       4.0       2.0     152.0  -37.84663   144.95681   
11652      3      16.1       3.0       1.0     726.0  -37.80279   145.19974   

       Propertycount  
7623          5533.0  
776           6795.0  
5684         14887.0  
12888         3280.0  
11652         4790.0

TypeError: 'Index' object is not callable